go/common/grpc/policy/watcher.go

```diff
@@ -2,35 +2,83 @@ package policy
 
 import (
 	"context"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
 
 	"github.com/oasisprotocol/oasis-core/go/common"
 	"github.com/oasisprotocol/oasis-core/go/common/accessctl"
 	"github.com/oasisprotocol/oasis-core/go/common/grpc"
 	"github.com/oasisprotocol/oasis-core/go/common/grpc/policy/api"
-	"github.com/oasisprotocol/oasis-core/go/common/pubsub"
+	"github.com/oasisprotocol/oasis-core/go/common/identity"
+	"github.com/oasisprotocol/oasis-core/go/common/logging"
+	"github.com/oasisprotocol/oasis-core/go/common/node"
+	sentry "github.com/oasisprotocol/oasis-core/go/sentry/api"
+	sentryClient "github.com/oasisprotocol/oasis-core/go/sentry/client"
 )
 
 var _ api.PolicyWatcher = (*policyWatcher)(nil)
 
 type policyWatcher struct {
-	policyNotifier *pubsub.Broker
+	ctx context.Context
+
+	sentryAddrs   []node.TLSAddress
+	id            *identity.Identity
+	sentryClients []*sentryClient.Client
+
+	logger *logging.Logger
 }
 
 func (c *policyWatcher) PolicyUpdated(service grpc.ServiceName, accessPolicies map[common.Namespace]accessctl.Policy) {
-	c.policyNotifier.Broadcast(api.ServicePolicies{Service: service, AccessPolicies: accessPolicies})
-}
+	// Notify the sentry nodes of the new policy.
+	for idx, addr := range c.sentryAddrs {
+		pushPolicies := func() error {
+			var client *sentryClient.Client
+			var err error
 
-func (c *policyWatcher) WatchPolicies(ctx context.Context) (<-chan api.ServicePolicies, pubsub.ClosableSubscription, error) {
-	typedCh := make(chan api.ServicePolicies)
-	sub := c.policyNotifier.Subscribe()
-	sub.Unwrap(typedCh)
+			if c.sentryClients[idx] != nil {
+				client = c.sentryClients[idx]
+			} else {
+				client, err = sentryClient.New(addr, c.id)
```

Comment:
Is there any reason why we wouldn't initialize all the clients in the constructor and only reference them here instead of creating new connections each time?

Reply:
Yes -- also, it's not creating new connections each time, they're cached.
Basically, this code does lazy connecting and also handles reconnecting if the connection is interrupted (e.g. when certs are rotated).  If we connected only in the constructor, then it would stop working after a while or it might not work at all in the beginning, since the sentry node might not be up yet.

Reply:
I overlooked that, looks ok then.